test/integration/scheduler/local-pv-neg-affinity_test.go

```diff
@@ -0,0 +1,335 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+// This file tests the VolumeScheduling feature.
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/controller/volume/persistentvolume"
+	"k8s.io/kubernetes/plugin/pkg/scheduler"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/factory"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	affinityLabelKey = "kubernetes.io/hostname"
+)
+
+func TestLocalPVNegativeAffinity(t *testing.T) {
+	config := setupNodes(t, "volume-scheduling", 3)
+	defer config.teardown()
+
+	pv := makeHostBoundPV(t, "local-pv", classImmediate, "", "", "node-1")
+	pvc := makePVC("local-pvc", config.ns, &classImmediate, "")
+
+	// Create PV
+	if _, err := config.client.CoreV1().PersistentVolumes().Create(pv); err != nil {
+		t.Fatalf("Failed to create PersistentVolume %q: %v", pv.Name, err)
+	}
+
+	// Create PVC
+	if _, err := config.client.CoreV1().PersistentVolumeClaims(config.ns).Create(pvc); err != nil {
+		t.Fatalf("Failed to create PersistentVolumeClaim %q: %v", pvc.Name, err)
+	}
+
+	nodeMarkers := []interface{}{
+		markNodeAffinity,
+		markNodeSelector,
+	}
+	for i := 0; i < len(nodeMarkers); i++ {
+		podName := "local-pod-" + strconv.Itoa(i+1)
+		pod := makePod(podName, config.ns, []string{"local-pvc"})
+		nodeMarkers[i].(func(*v1.Pod, string))(pod, "node-2")
+		// Create Pod
+		if _, err := config.client.CoreV1().Pods(config.ns).Create(pod); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+		// Give time to shceduler to attempt to schedule pod
+		if err := waitForPodToSchedule(config.client, pod); err == nil {
+			t.Errorf("Failed as Pod %s was scheduled sucessfully but expected to fail", pod.Name)
+		}
+		selector := fields.Set{
+			"involvedObject.kind":      "Pod",
+			"involvedObject.name":      podName,
+			"involvedObject.namespace": config.ns,
+			"reason":                   "FailedScheduling",
+		}.AsSelector().String()
+		options := metav1.ListOptions{FieldSelector: selector}
+		events, err := config.client.CoreV1().Events(config.ns).List(options)
```

Comment:
Why are you relying on events (which is subject to change) here.  If the data is not encoded in pod status then that is a bug.

Reply:
@timothysc thanks for your review. I do not think POD status provides very granular failure. In this test case we test a very specific scenario and scheduler reaction on it. @msau42 Appreciate your input here.

Reply:
The pod status would just be "pending", which I think is too generic of an error that we want to be testing.

Reply:
There are errors should be encoded into the status.message - https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#podstatus-v1-core  that indicate the failure.  I'm ok with the event for now, but please add a // TODO about reevaluating or removing the event requirement on the test.